sql/2020/03_Markup/pages_almanac_by_device_and_html_lang.sql

```diff
@@ -0,0 +1,43 @@
+#standardSQL
+# page almanac metrics grouped by device and html lang
+
+# helper to create percent fields
+CREATE TEMP FUNCTION AS_PERCENT (freq FLOAT64, total FLOAT64) RETURNS FLOAT64 AS (
+  ROUND(SAFE_DIVIDE(freq, total), 4)
+);
+
+# returns all the data we need from _almanac
+CREATE TEMPORARY FUNCTION get_almanac_html_lang(almanac_string STRING)
+RETURNS STRING LANGUAGE js AS '''
+try {
+    var almanac = JSON.parse(almanac_string); 
+
+    if (Array.isArray(almanac) || typeof almanac != 'object') return '';
+
+    if (almanac.html_node && almanac.html_node.lang) {
+      return almanac.html_node.lang.trim().toLowerCase();
+    }
+
+} catch (e) {}
+return '';
+''';
+
+SELECT
+  client,
+  COUNT(0) AS total,
+  almanac_html_lang as html_lang,
+
+  AS_PERCENT(COUNT(0), SUM(COUNT(0)) OVER (PARTITION BY client)) AS pct_m405
```

Comment:
Should the denominator be the number of pages? Or is the sum always equal to that?

Reply:
I'm not great on how the totalling works. I think this is fine because there is always one value per page?

Now I have to re-add my test code. And then I forgot to change the comments, just did a live run.

Test run shows total of freq for a client equals the number of pages (10k) indicating it should be fine.  For the live run freq for both clients totalled 11,941,282 which also sounds about right.

Added an ORDER BY for next time but was easy to do that in the sheet.

Reply:
Seems Japanese beats Spanish